Mosaic image fusion blogpost (Volker Hilsenstein) (#108)
dask/dask-blog: _posts/2021-08-29-mosaic-fusion.md

---
layout: post
title: Mosaic Image Fusion
author: Volker Hilsenstein, Marvin Albert, and Genevieve Buckley
tags: [life science]
theme: twitter
---
{% include JB/setup %}

## Executive Summary

This blogpost shows a case study, where a researcher uses Dask for mosaic image fusion.
Mosaic image fusion is when you combine multiple smaller images taken at known locations and stitch them together into a single image with a very large field of view. Full code examples are available on GitHub from the `DaskFusion` repository: https://github.com/VolkerH/DaskFusion

> This work was done by Volker Hilsenstein, in conjunction with Marvin Albert. Genevieve Buckley and Volker Hilsenstein wrote the blogpost. Volker Hilsenstein is a scientific software developer at [EMBL in Theodore Alexandrov's lab](https://www.embl.org/groups/alexandrov/) with a focus on spatial metabolomics and bio-image analysis.

Author: ```suggestion
This work was done by Volker Hilsenstein, in conjunction with Marvin Albert. Genevieve Buckley and Volker Hilsenstein wrote the blogpost. Volker Hilsenstein is a scientific software developer at [EMBL in Theodore Alexandrov's lab](https://www.embl.org/groups/alexandrov/) with a focus on spatial metabolomics and bio-image analysis. The sample has been prepared and imaged by Mohammed Shahraz from the Alexandrov lab at EMBL Heidelberg.
```

## The problem

### Image mosaicing in microscopy

In optical microscopy, a single field of view captured with a 20x objective typically
has a diagonal on the order of a few 100 um (exact dimensions depend on other
parts of the optical system, including the size of the camera chip). A typical 
sample slide has a size of 25mm * 70mm. 
So when imaging a whole slide, one has to acquire hundreds of images, typically
with some overlap between individual tiles.  With increasing magnifcation,
the required number of images increases accordingly. 

To obtain an overview one has to fuse this large number of individual
image tiles into a large mosaic image. Herer, we assume that the inforamtion requiring for 
positioning and alignment of the individual image tiles is already available. In our case,
this information is available based on the the metadata recorded by the microscope, however this
information could also be from  previous registration step that matches image features 
in the overlapping areas.
 

## The solution

The array that can hold resulting mosaic image will often have a size that is too large 

> to fit in RAM, therefore we will use dask arrays and the `map_block` function to enable 

Author: ```suggestion
to fit in RAM, therefore we will use dask arrays and the [`map_blocks`](https://docs.dask.org/en/latest/generated/dask.array.map_blocks.html) function to enable 
```

Typically whenever we want to join dask arrays, we use [Stack, Concatenate, and Block](https://docs.dask.org/en/latest/array-stack.html). However, these are not good tools for mosaic image fusion, because:

1. the image tiles will be be overlapping,
2. in the general case, individual image tiles may not be positioned on an exact grid and they can also have small rotations.

Using the `block_info` dictionary and creating some pseudo-code together, Volker managed to implement mosaic fusion with `dask-image`. From the [Dask documentaion](https://docs.dask.org/en/latest/generated/dask.array.map_blocks.html?highlight=block_info#dask.array.map_blocks):
> Your block function gets information about where it is in the array by accepting a special `block_info` or `block_id` keyword argument.

The basic outline of the image mosiac workflow is as follows:
1. Read in image tiles acquired from the microscope.
2. Calculate the location of the image tiles, using information from the sample stage position at the time each image tile was acquired.
3. Use an affine transformation to place the image tiles in their proper location.
4. Fuse the combined images into a single result.

### Results

> Using the dask-based image fusion method, we could achieve a speed up from several hours to tens of minutes compared to 

Author: ```suggestion
Using the Dask-based image fusion method, we could achieve a speed up from several hours to tens of minutes compared to 
```

a previous method based on an ImageJ plugin for datasets with many image tiles (~500-1000 tiles) on similar workstation hardware.

> Due to dask's ability to handle data out-of-core and chunked array storage using zarr it is also possible to run the 

Author: ```suggestion
Due to Dask's ability to handle data out-of-core and chunked array storage using zarr it is also possible to run the 
```

(very small nit-pick, since apparently there is a [style guide](https://marketing.dask.org/en/latest/refer.html) and we're supposed to spell Dask with a capital first letter)

fusion on hardware with limited RAM.

> The image below shows mosaic image fusion for a smaall examples with 63 image tiles, displayed in the napari image viewer.

Author: ```suggestion
The image below shows mosaic image fusion for a small example with 63 image tiles, displayed in the napari image viewer.
```

Author: Small typo

![Mosaic fusion images in the napari image viewer](/images/mosaic-fusion/NapariMosaics.png)

> And here is an animation of whole slide mosaic fusion.

Author: Does this line need to be rephrased? 

> just letting you know that some of the images you included (I assume from the napari Zulip) for illustration were not showing the Dask Fusion. Those were based on placing the individual tiles (heavily downsampled) on the napari canvas using affine translations. However, in contrast to using dask fusion that approach does not scale to 100s of tiles unless you heavily downsamples.

Author: Rephrased now.

![Animation of whole slide mosaic fusion images](/images/mosaic-fusion/Lama_whole_slide.gif)

### Code

Code relatiing to this mosaic image fusion project can be found in the `DaskFusion` GitHub repository here: https://github.com/VolkerH/DaskFusion

> There is a self-contained example available in [this notebook](https://github.com/VolkerH/DaskFusion/blob/main/Load_Mosaic.ipynb), which downloads reduced-size example data to demonstrate the process.

Author: ```suggestion
There is a self-contained example available in [this notebook](https://github.com/VolkerH/DaskFusion/blob/main/DaskFusion_Example.ipynb), which downloads reduced-size example data to demonstrate the process.
```

Notebook has been renamed (not yet merged, but we can assume it will be by the time this is published)

## What's next?

Currently, the DaskFusion code is a proof of concept for single-channel 2D images and simple maximum projection for blending the tiles in overlapping areas, it is not production code.
However, the same principle can be used for fusing multi-channel image volumes,
such as from Light-Sheet data if the tile chunk intersection calculation is extended to higher-dimensional arrays.
Such even larger datasets will benefit even more from leveraging dask,
as the processing can be distributed across multiple nodes of a HPC cluster using [dask jobqueue](http://jobqueue.dask.org/en/latest/).

### Also see
Marvin's lightning talk on multi-view image fusion is up online now: https://www.youtube.com/watch?v=YIblUvonMvo&list=PLJ0vO2F_f6OBAY6hjRHM_mIQ9yh32mWr0&index=10

The GitHub repository [MVRegFus](https://github.com/m-albert/MVRegFus) that Marvin talks about in the video is available here: https://github.com/m-albert/MVRegFus

## Acknowledgements

> The sample images are courtesy of M. Shahraz, EMBL Alexandrov Lab.

Author: Let's put all of the acknowledgements about who did what in one place. I don't mind where, either at the top where the others currently are, or we could move it all down to this section instead.

---
layout: post
title: Mosaic Image Fusion
author: Volker Hilsenstein and Genevieve Buckley
tags: [life science]
theme: twitter
---
{% include JB/setup %}

## Executive Summary

> This blogpost shows a case study, where a researcher uses Dask for mosaic image fusion - stitching multiple smaller images together seamlessly into a very large field of view. Full code examples are available on GitHub from the `DaskFusion` repository: https://github.com/VolkerH/DaskFusion

Author: This repository is currently private.

Reviewer: I know, I was not sure whether the code would be included in the blogpost repo (in which case I would just included the relevant bits).

Author: I think it's probably more useful to link to a notebook or repository (that way people don't have to copy-paste bits together). 

Including small snippets of code in the blogpost might also be useful, if it's helpful to have an illustrative example?

Author: Note: The repository https://github.com/VolkerH/DaskFusion is public now

> This work was done by Volker Hilsenstein, a scientific software developer at EMBL in Theodore Alexandrov's lab. Special thanks go to Marvin Albert for valuable discussions and advice. Genevieve Buckley wrote the blogpost.

Author: Also needs a thank you to the data owners here.

## The problem

> Volker works on [spatial metabolomics](https://www.ebi.ac.uk/training/online/courses/metabolomics-introduction/what-is/) and bio-image analysis.

Author: Maybe @VolkerH could add a sentence here describing what sort of samples his lab looks at, etc.

However, the researchers discovered that in practice for large numbers of tiles (several hundred) there was an unfortunate scaling behaviour, estimating roughly `O(n^2)` complexity or worse. (If you want to know more about what that means and why it's a bad thing, see this article on [Learning Big O Notation With O(n) Complexity](https://dzone.com/articles/learning-big-o-notation-with-on-complexity)). This is what prompted them to explore alternative options, and finally settle on using [Dask](https://dask.org/) to manage the image fusion, and the [napari image viewer](https://napari.org/) to display the results.

## The solution

Typically whenever we want to join dask arrays, we use [Stack, Concatenate, and Block](https://docs.dask.org/en/latest/array-stack.html). However, these are not good tools for mosaic image fusion, because:

1. There needs to be some overlap at the edges of each image tile, and
2. While the position of the image tile is known based on the sample stage motor coordinates, there is a small amount of uncertainty in this measurement.

> 

Author: Text here needs finishing. Perhaps describe the basic workflow using Dask?

Author: Basic outline:
* image data from microscope
* coordinate data, indicating the sample stage position at the time the image tile was acquired
* affine transformation to place the image tile in its proper position (maybe using new dask-image function? double check utility function scripts for details)
* fusion

Marvin's lightning talk on multi-view image fusion is up online now: https://www.youtube.com/watch?v=YIblUvonMvo&list=PLJ0vO2F_f6OBAY6hjRHM_mIQ9yh32mWr0&index=10

[MVRegFus](https://github.com/m-albert/MVRegFus)

`block_info` dictionary and creating some
pseudo-code together, I managed to implement mosaic fusion in Dask Image.

From the dask documentaion:
> Your block function gets information about where it is in the array by accepting a special `block_info` or `block_id` keyword argument.

### Results

> Using the new method with Dask, a considerable speed-up with this implementation has been observed. Previously, for data with 600 image tiles it was taking several hours to complete the image fusion. Now with Dask, image fusion is complete in minutes instead of hours.

Author: This is pretty good, but can we be any more specific about the speedup than just "minutes instead of hours"?

The image below shows mosaic image fusion with 63 image tiles, displayed in the napari image viewer.

![Mosaic fusion images in the napari image viewer](/images/mosaic-fusion/NapariMosaics.png)

And here is an animation of whole slide mosaic fusion.

![Animation of whole slide mosaic fusion images](/images/mosaic-fusion/Lama_whole_slide.gif)

### Code

Code relatiing to this mosaic image fusion project can be found in the `DaskFusion` GitHub repository here: https://github.com/VolkerH/DaskFusion

There is a self-contained example available in [this notebook](https://github.com/VolkerH/DaskFusion/blob/main/Load_Mosaic.ipynb), which downloads reduced-size example data to demonstrate the process.

## What's next?

> Volker says, "If anyone is keen this could be amended for 3D and use Big Stitcher Project files as input". If that's something you might be interested in, get in touch!

Author: How should people get in touch? Any preferred contact method we should put here?

---
layout: post
title: Mosaic Image Fusion

> author: Volker Hilsenstein, Marvin Albert, and Genevieve Buckley

Author: Hi @m-albert - just making you aware that you're listed as a co-author of this blogpost, Volker tells me your contributions to the project were substantial.

tags: [life science]
theme: twitter
---
{% include JB/setup %}

## Executive Summary

This blogpost shows a case study, where a researcher uses Dask for mosaic image fusion - stitching multiple smaller images together seamlessly into a very large field of view. Full code examples are available on GitHub from the `DaskFusion` repository: https://github.com/VolkerH/DaskFusion

This work was done by Volker Hilsenstein, in conjunction with Marvin Albert. Genevieve Buckley wrote the blogpost.

> Volker Hilsenstein is a scientific software developer at EMBL in Theodore Alexandrov's lab.

Author: We need to write about who provided the data here too.

Reviewer: Yes, I was actually thinking I might take some images of something nicer specifically for this purpose or some calibration slide.

Author: I am slightly against this idea, mostly because I know you are quite busy and I don't want this to fall into the trap described by the saying "don't let perfect be the enemy of good" (i.e. don't try to make things perfect and never quite finish, better to put slightly less good things out more frequently).
You'll have a much better idea of how much extra effort that would actually be, though

To obtain an overview one has to fuse this large number of individual
image tiles into a large mosaic image. Herer, we assume that the inforamtion requiring for 
positioning and alignment of the individual image tiles is already available. In our case,
this information is available based on the the metadata recorded by the microscope, however this
information could also be from  previous registration step that matches image features 
in the overlapping areas.
 

## The solution

The array that can hold resulting mosaic image will often have a size that is too large 
to fit in RAM, therefore we will use dask arrays and the `map_block` function to enable 
out-of-core processing. As an added benefit, we will get parallel processing for free
which speeds up the fusion process.

Typically whenever we want to join dask arrays, we use [Stack, Concatenate, and Block](https://docs.dask.org/en/latest/array-stack.html). However, these are not good tools for mosaic image fusion, because:

1. the image tiles will be be overlapping,
2. in the general case, individual image tiles may not be positioned on an exact grid and they can also have small rotations.

> `block_info` dictionary and creating some

Author: Incomplete sentence here
